HEAD request on websocket throws ugly 500

Tripped over this by complete coincidence, but when you do a `HEAD` request to a websocket endpoint, you get a ugly internal server error (see stacktrace at the bottom).

I started poking around, but this is the only case I could come up with that would break, so I'm guessing this is the automagic `GET`s-also-serve-`HEAD`s code messing things up here, probably failing to return a `WebSocket.reject response`.
Other things I tried, but were handled perfectly fine:
- normal `GET` request to the websocket endpoint, without websocket headers (`400 bad request`, expecting this from the `HEAD` request too)
- other HTTP method requests (like `DELETE`, `OPTION`, `PUT`) to the websocket endpoint (`404 not found`)
- `HEAD` request to non websocket endpoints (`200 OK`)

It would be nice to get a clean response instead of a nasty stacktrace, even though it is unlikely that someone will run in to this.

```
[ERROR] from application in New I/O worker #19 - ! @6mpedf1kb - Internal server error, for (HEAD) [/news/stream] -> play.api.http.HttpErrorHandlerExceptions$$anon$1: Execution exception[[MatchError: Some(WebSocket(<function1>)) (of class scala.Some)]]
    at play.api.http.HttpErrorHandlerExceptions$.throwableToUsefulException(HttpErrorHandler.scala:265) ~[play_2.11-2.4.2.jar:2.4.2]
    at play.api.http.DefaultHttpErrorHandler.onServerError(HttpErrorHandler.scala:191) ~[play_2.11-2.4.2.jar:2.4.2]
    at play.core.server.Server$class.logExceptionAndGetResult$1(Server.scala:50) [play-server_2.11-2.4.2.jar:2.4.2]
    at play.core.server.Server$$anonfun$getHandlerFor$4.apply(Server.scala:59) [play-server_2.11-2.4.2.jar:2.4.2]
    at play.core.server.Server$$anonfun$getHandlerFor$4.apply(Server.scala:57) [play-server_2.11-2.4.2.jar:2.4.2]
    at scala.util.Either$RightProjection.flatMap(Either.scala:522) [scala-library-2.11.7.jar:na]
    at play.core.server.Server$class.getHandlerFor(Server.scala:57) [play-server_2.11-2.4.2.jar:2.4.2]
    at play.core.server.NettyServer.getHandlerFor(NettyServer.scala:33) [play-netty-server_2.11-2.4.2.jar:2.4.2]
    at play.core.server.netty.PlayDefaultUpstreamHandler$$anonfun$8.apply(PlayDefaultUpstreamHandler.scala:139) [play-netty-server_2.11-2.4.2.jar:2.4.2]
    at play.core.server.netty.PlayDefaultUpstreamHandler$$anonfun$8.apply(PlayDefaultUpstreamHandler.scala:139) [play-netty-server_2.11-2.4.2.jar:2.4.2]
    at scala.util.Either.fold(Either.scala:99) [scala-library-2.11.7.jar:na]
    at play.core.server.netty.PlayDefaultUpstreamHandler.messageReceived(PlayDefaultUpstreamHandler.scala:127) [play-netty-server_2.11-2.4.2.jar:2.4.2]
    at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70) [netty-3.10.3.Final.jar:na]
    at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564) [netty-3.10.3.Final.jar:na]
    at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791) [netty-3.10.3.Final.jar:na]
    at com.typesafe.netty.http.pipelining.HttpPipeliningHandler.messageReceived(HttpPipeliningHandler.java:62) [netty-http-pipelining-1.1.4.jar:na]
    at org.jboss.netty.channel.SimpleChannelHandler.handleUpstream(SimpleChannelHandler.java:88) [netty-3.10.3.Final.jar:na]
    at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564) [netty-3.10.3.Final.jar:na]
    at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791) [netty-3.10.3.Final.jar:na]
    at org.jboss.netty.handler.codec.http.HttpContentDecoder.messageReceived(HttpContentDecoder.java:108) [netty-3.10.3.Final.jar:na]
    at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70) [netty-3.10.3.Final.jar:na]
    at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564) [netty-3.10.3.Final.jar:na]
    at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791) [netty-3.10.3.Final.jar:na]
    at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296) [netty-3.10.3.Final.jar:na]
    at org.jboss.netty.handler.codec.frame.FrameDecoder.unfoldAndFireMessageReceived(FrameDecoder.java:459) [netty-3.10.3.Final.jar:na]
    at org.jboss.netty.handler.codec.replay.ReplayingDecoder.callDecode(ReplayingDecoder.java:536) [netty-3.10.3.Final.jar:na]
    at org.jboss.netty.handler.codec.replay.ReplayingDecoder.messageReceived(ReplayingDecoder.java:435) [netty-3.10.3.Final.jar:na]
    at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70) [netty-3.10.3.Final.jar:na]
    at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564) [netty-3.10.3.Final.jar:na]
    at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:559) [netty-3.10.3.Final.jar:na]
    at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:268) [netty-3.10.3.Final.jar:na]
    at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:255) [netty-3.10.3.Final.jar:na]
    at org.jboss.netty.channel.socket.nio.NioWorker.read(NioWorker.java:88) [netty-3.10.3.Final.jar:na]
    at org.jboss.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:108) [netty-3.10.3.Final.jar:na]
    at org.jboss.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:337) [netty-3.10.3.Final.jar:na]
    at org.jboss.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:89) [netty-3.10.3.Final.jar:na]
    at org.jboss.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178) [netty-3.10.3.Final.jar:na]
    at org.jboss.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108) [netty-3.10.3.Final.jar:na]
    at org.jboss.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42) [netty-3.10.3.Final.jar:na]
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_51]
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_51]
    at java.lang.Thread.run(Thread.java:745) [na:1.8.0_51]
Caused by: scala.MatchError: Some(WebSocket(<function1>)) (of class scala.Some)
    at play.api.GlobalSettings$$anonfun$3.apply(GlobalSettings.scala:119) ~[play_2.11-2.4.2.jar:2.4.2]
    at play.api.GlobalSettings$$anonfun$3.apply(GlobalSettings.scala:113) ~[play_2.11-2.4.2.jar:2.4.2]
    at scala.Option.getOrElse(Option.scala:121) ~[scala-library-2.11.7.jar:na]
    at play.api.GlobalSettings$class.onRequestReceived(GlobalSettings.scala:113) ~[play_2.11-2.4.2.jar:2.4.2]
    at play.core.j.JavaGlobalSettingsAdapter.onRequestReceived(JavaGlobalSettingsAdapter.scala:14) ~[play_2.11-2.4.2.jar:2.4.2]
    at play.api.http.GlobalSettingsHttpRequestHandler.handlerForRequest(HttpRequestHandler.scala:183) ~[play_2.11-2.4.2.jar:2.4.2]
    at play.core.server.Server$$anonfun$sendHandler$1$1.apply(Server.scala:38) ~[play-server_2.11-2.4.2.jar:2.4.2]
    at play.core.server.Server$$anonfun$sendHandler$1$1.apply(Server.scala:37) ~[play-server_2.11-2.4.2.jar:2.4.2]
    at scala.util.Success$$anonfun$map$1.apply(Try.scala:237) ~[scala-library-2.11.7.jar:na]
    at scala.util.Try$.apply(Try.scala:192) ~[scala-library-2.11.7.jar:na]
    at scala.util.Success.map(Try.scala:237) ~[scala-library-2.11.7.jar:na]
    at play.core.server.Server$class.sendHandler$1(Server.scala:37) [play-server_2.11-2.4.2.jar:2.4.2]
    at play.core.server.Server$$anonfun$getHandlerFor$4.apply(Server.scala:58) [play-server_2.11-2.4.2.jar:2.4.2]
    ... 38 common frames omitted
```

We see it in production:

```
! @73897i09l - Internal server error, for (HEAD) [/api/feed/socket] ->
 
play.api.http.HttpErrorHandlerExceptions$$anon$1: Execution exception[[MatchError: Some(WebSocket(<function1>)) (of class scala.Some)]]
        at play.api.http.HttpErrorHandlerExceptions$.throwableToUsefulException(HttpErrorHandler.scala:265) ~[com.typesafe.play.play_2.11-2.4.8.jar:2.4.8]
        at play.api.http.DefaultHttpErrorHandler.onServerError(HttpErrorHandler.scala:191) ~[com.typesafe.play.play_2.11-2.4.8.jar:2.4.8]
        at play.core.server.Server$class.logExceptionAndGetResult$1(Server.scala:50) [com.typesafe.play.play-server_2.11-2.4.8.jar:2.4.8]
        at play.core.server.Server$$anonfun$getHandlerFor$4.apply(Server.scala:59) [com.typesafe.play.play-server_2.11-2.4.8.jar:2.4.8]
        at play.core.server.Server$$anonfun$getHandlerFor$4.apply(Server.scala:57) [com.typesafe.play.play-server_2.11-2.4.8.jar:2.4.8]
        at scala.util.Either$RightProjection.flatMap(Either.scala:522) [org.scala-lang.scala-library-2.11.8.jar:na]
        at play.core.server.Server$class.getHandlerFor(Server.scala:57) [com.typesafe.play.play-server_2.11-2.4.8.jar:2.4.8]
        at play.core.server.NettyServer.getHandlerFor(NettyServer.scala:37) [com.typesafe.play.play-netty-server_2.11-2.4.8.jar:2.4.8]
        at play.core.server.netty.PlayDefaultUpstreamHandler$$anonfun$8.apply(PlayDefaultUpstreamHandler.scala:143) [com.typesafe.play.play-netty-server_2.11-2.4.8.jar:2.4.8]
        at play.core.server.netty.PlayDefaultUpstreamHandler$$anonfun$8.apply(PlayDefaultUpstreamHandler.scala:143) [com.typesafe.play.play-netty-server_2.11-2.4.8.jar:2.4.8]
        at scala.util.Either.fold(Either.scala:99) [org.scala-lang.scala-library-2.11.8.jar:na]
        at play.core.server.netty.PlayDefaultUpstreamHandler.messageReceived(PlayDefaultUpstreamHandler.scala:131) [com.typesafe.play.play-netty-server_2.11-2.4.8.jar:2.4.8]
        at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70) [io.netty.netty-3.10.5.Final.jar:na]
        at org.jboss.netty.handler.timeout.IdleStateAwareChannelUpstreamHandler.handleUpstream(IdleStateAwareChannelUpstreamHandler.java:36) [io.netty.netty-3.10.5.Final.jar:na]
        at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564) [io.netty.netty-3.10.5.Final.jar:na]
        at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791) [io.netty.netty-3.10.5.Final.jar:na]
        at com.typesafe.netty.http.pipelining.HttpPipeliningHandler.messageReceived(HttpPipeliningHandler.java:62) [com.typesafe.netty.netty-http-pipelining-1.1.4.jar:na]
        at org.jboss.netty.channel.SimpleChannelHandler.handleUpstream(SimpleChannelHandler.java:88) [io.netty.netty-3.10.5.Final.jar:na]
        at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564) [io.netty.netty-3.10.5.Final.jar:na]
        at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791) [io.netty.netty-3.10.5.Final.jar:na]
        at org.jboss.netty.handler.codec.http.HttpContentDecoder.messageReceived(HttpContentDecoder.java:108) [io.netty.netty-3.10.5.Final.jar:na]
        at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70) [io.netty.netty-3.10.5.Final.jar:na]
        at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564) [io.netty.netty-3.10.5.Final.jar:na]
        at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791) [io.netty.netty-3.10.5.Final.jar:na]
        at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296) [io.netty.netty-3.10.5.Final.jar:na]
        at org.jboss.netty.handler.codec.frame.FrameDecoder.unfoldAndFireMessageReceived(FrameDecoder.java:459) [io.netty.netty-3.10.5.Final.jar:na]
        at org.jboss.netty.handler.codec.replay.ReplayingDecoder.callDecode(ReplayingDecoder.java:536) [io.netty.netty-3.10.5.Final.jar:na]
        at org.jboss.netty.handler.codec.replay.ReplayingDecoder.messageReceived(ReplayingDecoder.java:435) [io.netty.netty-3.10.5.Final.jar:na]
        at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70) [io.netty.netty-3.10.5.Final.jar:na]
        at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564) [io.netty.netty-3.10.5.Final.jar:na]
        at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:559) [io.netty.netty-3.10.5.Final.jar:na]
        at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:268) [io.netty.netty-3.10.5.Final.jar:na]
        at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:255) [io.netty.netty-3.10.5.Final.jar:na]
        at org.jboss.netty.channel.socket.nio.NioWorker.read(NioWorker.java:88) [io.netty.netty-3.10.5.Final.jar:na]
        at org.jboss.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:108) [io.netty.netty-3.10.5.Final.jar:na]
        at org.jboss.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:337) [io.netty.netty-3.10.5.Final.jar:na]
        at org.jboss.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:89) [io.netty.netty-3.10.5.Final.jar:na]
        at org.jboss.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178) [io.netty.netty-3.10.5.Final.jar:na]
        at org.jboss.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108) [io.netty.netty-3.10.5.Final.jar:na]
        at org.jboss.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42) [io.netty.netty-3.10.5.Final.jar:na]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_121]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_121]
        at java.lang.Thread.run(Thread.java:745) [na:1.8.0_121]
Caused by: scala.MatchError: Some(WebSocket(<function1>)) (of class scala.Some)
        at play.api.GlobalSettings$$anonfun$3.apply(GlobalSettings.scala:119) ~[com.typesafe.play.play_2.11-2.4.8.jar:2.4.8]
        at play.api.GlobalSettings$$anonfun$3.apply(GlobalSettings.scala:113) ~[com.typesafe.play.play_2.11-2.4.8.jar:2.4.8]
        at scala.Option.getOrElse(Option.scala:121) ~[org.scala-lang.scala-library-2.11.8.jar:na]
        at play.api.GlobalSettings$class.onRequestReceived(GlobalSettings.scala:113) ~[com.typesafe.play.play_2.11-2.4.8.jar:2.4.8]
        at play.api.DefaultGlobal$.onRequestReceived(GlobalSettings.scala:212) ~[com.typesafe.play.play_2.11-2.4.8.jar:2.4.8]
        at play.api.http.GlobalSettingsHttpRequestHandler.handlerForRequest(HttpRequestHandler.scala:183) ~[com.typesafe.play.play_2.11-2.4.8.jar:2.4.8]
        at play.core.server.Server$$anonfun$sendHandler$1$1.apply(Server.scala:38) ~[com.typesafe.play.play-server_2.11-2.4.8.jar:2.4.8]
        at play.core.server.Server$$anonfun$sendHandler$1$1.apply(Server.scala:37) ~[com.typesafe.play.play-server_2.11-2.4.8.jar:2.4.8]
        at scala.util.Success$$anonfun$map$1.apply(Try.scala:237) ~[org.scala-lang.scala-library-2.11.8.jar:na]
        at scala.util.Try$.apply(Try.scala:192) ~[org.scala-lang.scala-library-2.11.8.jar:na]
        at scala.util.Success.map(Try.scala:237) ~[org.scala-lang.scala-library-2.11.8.jar:na]
        at play.core.server.Server$class.sendHandler$1(Server.scala:37) [com.typesafe.play.play-server_2.11-2.4.8.jar:2.4.8]
        at play.core.server.Server$$anonfun$getHandlerFor$4.apply(Server.scala:58) [com.typesafe.play.play-server_2.11-2.4.8.jar:2.4.8]
        ... 39 common frames omitted
```

--- a/framework/src/play-akka-http-server/src/main/resources/reference.conf
+++ b/framework/src/play-akka-http-server/src/main/resources/reference.conf
@@ -27,7 +27,7 @@ play {
       # bodies from outgoing responses.
       # Note that, even when this setting is off the server will never send
       # out message bodies on responses to HEAD requests.
-      transparent-head-requests = on
+      transparent-head-requests = off
 
       # If this setting is empty the server only accepts requests that carry a
       # non-empty `Host` header. Otherwise it responds with `400 Bad Request`.

--- a/framework/src/play-akka-http-server/src/main/scala/play/core/server/AkkaHttpServer.scala
+++ b/framework/src/play-akka-http-server/src/main/scala/play/core/server/AkkaHttpServer.scala
@@ -26,7 +26,6 @@ import play.api.libs.streams.Accumulator
 import play.api.mvc._
 import play.api.routing.Router
 import play.core.{ ApplicationProvider, DefaultWebCommands, SourceMapper, WebCommands }
-import play.core.server._
 import play.core.server.akkahttp.{ AkkaModelConversion, HttpRequestDecoder }
 import play.core.server.common.{ ForwardedHeaderHandler, ServerResultUtils }
 import play.core.server.ssl.ServerSSLEngine

--- a/framework/src/play-integration-test/src/test/scala/play/it/action/HeadActionSpec.scala
+++ b/framework/src/play-integration-test/src/test/scala/play/it/action/HeadActionSpec.scala
@@ -7,13 +7,11 @@ package play.it.action
 import akka.stream.scaladsl.Source
 import play.shaded.ahc.io.netty.handler.codec.http.HttpHeaders
 import org.specs2.mutable.Specification
-import play.api.Play
 import play.api.http.HeaderNames._
 import play.api.http.Status._
 import play.api.libs.ws.{ WSClient, WSResponse }
 import play.api.mvc._
 import play.api.routing.Router.Routes
-import play.api.routing.Router.Tags._
 import play.api.routing.sird._
 import play.api.test._
 import play.core.server.Server
@@ -24,17 +22,22 @@ import scala.concurrent.ExecutionContext.Implicits.global
 import play.shaded.ahc.org.asynchttpclient.netty.NettyResponse
 import play.api.libs.typedmap.TypedKey
 
+import scala.concurrent.Future
+
 class NettyHeadActionSpec extends HeadActionSpec with NettyIntegrationSpecification
 class AkkaHttpHeadActionSpec extends HeadActionSpec with AkkaHttpIntegrationSpecification
 
 trait HeadActionSpec extends Specification with FutureAwaits with DefaultAwaitTimeout with ServerIntegrationSpecification {
 
-  private def route(verb: String, path: String)(handler: EssentialAction): PartialFunction[(String, String), Handler] = {
-    case (v, p) if v == verb && p == path => handler
-  }
   sequential
 
   "HEAD requests" should {
+
+    def webSocketResponse(implicit Action: DefaultActionBuilder): Routes = {
+      case GET(p"/ws") => WebSocket.acceptOrResult[String, String] { request =>
+        Future.successful(Left(Results.Forbidden))
+      }
+    }
 
     def chunkedResponse(implicit Action: DefaultActionBuilder): Routes = {
       case GET(p"/chunked") =>
@@ -51,6 +54,7 @@ trait HeadActionSpec extends Specification with FutureAwaits with DefaultAwaitTi
         .orElse(delete) // DELETE /delete
         .orElse(stream) // GET /stream/0
         .orElse(chunkedResponse) // GET /chunked
+        .orElse(webSocketResponse) // GET /ws
 
     def withServer[T](block: WSClient => T): T = {
       // Routes from HttpBinApplication
@@ -65,6 +69,11 @@ trait HeadActionSpec extends Specification with FutureAwaits with DefaultAwaitTi
       } { implicit port =>
         WsTestClient.withClient(block)
       }
+    }
+
+    "return 400 in response to a HEAD in a WebSocket handler" in withServer { client =>
+      val result = await(client.url("/ws").head())
+      result.status must_== BAD_REQUEST
     }
 
     "return 200 in response to a URL with a GET handler" in withServer { client =>

--- a/framework/src/play/src/main/scala/play/api/http/HttpRequestHandler.scala
+++ b/framework/src/play/src/main/scala/play/api/http/HttpRequestHandler.scala
@@ -109,6 +109,11 @@ class DefaultHttpRequestHandler(router: Router, errorHandler: HttpErrorHandler, 
   }
 
   override def handlerForRequest(request: RequestHeader): (RequestHeader, Handler) = {
+
+    def handleWithStatus(status: Int) = ActionBuilder.ignoringBody.async(BodyParsers.utils.empty)(req =>
+      errorHandler.onClientError(req, status)
+    )
+
     /**
      * Call the router to get the handler, but with a couple of types of fallback.
      * First, if a HEAD request isn't explicitly routed try routing it as a GET
@@ -116,19 +121,31 @@ class DefaultHttpRequestHandler(router: Router, errorHandler: HttpErrorHandler, 
      * error.
      */
     def routeWithFallback(request: RequestHeader): Handler = {
-      routeRequest(request) match {
-        case Some(handler) => handler
-        // We automatically permit HEAD requests against any GETs without the need to
-        // add an explicit mapping in Routes. Since we couldn't route the HEAD request,
-        // try to get a Handler for the equivalent GET request instead. Note: the handler
-        // returned will still be passed a HEAD request when it is actually evaluated.
-        case None if request.method == HttpVerbs.HEAD =>
-          routeWithFallback(request.withMethod(HttpVerbs.GET))
-        case None =>
-          // An Action for a 404 error
-          ActionBuilder.ignoringBody.async(BodyParsers.utils.empty)(req =>
-            errorHandler.onClientError(req, NOT_FOUND)
-          )
+      routeRequest(request).getOrElse {
+        request.method match {
+          // We automatically permit HEAD requests against any GETs without the need to
+          // add an explicit mapping in Routes. Since we couldn't route the HEAD request,
+          // try to get a Handler for the equivalent GET request instead. Notes:
+          // 1. The handler returned will still be passed a HEAD request when it is
+          //    actually evaluated.
+          // 2. When the endpoint is to a WebSocket connection, the handler returned
+          //    will result in a Bad Request. That is because, while we can translate
+          //    GET requests to HEAD, we can't do that for WebSockets, since there is
+          //    no way (or reason) to Upgrade the connection. For more information see
+          //    https://tools.ietf.org/html/rfc6455#section-1.3
+          case HttpVerbs.HEAD => {
+            routeRequest(request.withMethod(HttpVerbs.GET)) match {
+              case Some(handler: Handler) => handler match {
+                case ws: WebSocket => handleWithStatus(BAD_REQUEST)
+                case _ => handler
+              }
+              case None => handleWithStatus(NOT_FOUND)
+            }
+          }
+          case _ =>
+            // An Action for a 404 error
+            handleWithStatus(NOT_FOUND)
+        }
       }
     }
 

